### DISCUSSION TITLE
Function not starting orchestrator and causing SessionLockLostException

### DISCUSSION THREAD

**Discussion by adeliab** [Azure/azure-functions-durable-extension#1985] [Q&A]
I have a service bus triggered function - it listens to a subscription that is session based. This function then triggers an orchestration function.

The trigger function logic is the following:
- Start a new orchestrator
- It calls starter.GetStatusAsync() with retry interval for 1 minute, checking whether the orchestration has finished (not running or pending)
- If the orchestration is not finished in a minute, it calls messageReceiver.AbandonAsync()

At a certain point, the function in all our non-production environments fails to execute the orchestrator. The orchestrator record is created in the instance table, but it keeps status pending for a long time.

This cause the AbandonAsync statement to get triggered - and when this is triggered, we receive SessionLockLostException.

`Exception while executing function: xxxxx ---> Microsoft.Azure.ServiceBus.SessionLockLostException : Session lock lost. Accept a new session
   at async Microsoft.Azure.ServiceBus.Core.MessageReceiver.DisposeMessagesAsync(IEnumerable1 lockTokens,Outcome outcome)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at async Microsoft.Azure.ServiceBus.RetryPolicy.RunOperation(Func1 operation,TimeSpan operationTimeout)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at async Microsoft.Azure.ServiceBus.RetryPolicy.RunOperation(Func1 operation,TimeSpan operationTimeout)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at async Microsoft.Azure.ServiceBus.Core.MessageReceiver.AbandonAsync(String lockToken,IDictionary2 propertiesToModify)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at async DeviceManagement.xxxxx.Run(IDurableOrchestrationClient starter,Message message,MessageReceiver messageReceiver) at D:\a\1\s\xxxxx
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at async Microsoft.Azure.WebJobs.Host.Executors.VoidTaskMethodInvoker2.InvokeAsync[TReflected,TReturnType](TReflected instance,Object[] arguments) at C:\projects\azure-webjobs-sdk-rqm4t\src\Microsoft.Azure.WebJobs.Host\Executors\VoidTaskMethodInvoker.cs : 20
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at async Microsoft.Azure.WebJobs.Host.Executors.FunctionInvoker2.InvokeAsync[TReflected,TReturnValue](Object instance,Object[] arguments) at C:\projects\azure-webjobs-sdk-rqm4t\src\Microsoft.Azure.WebJobs.Host\Executors\FunctionInvoker.cs : 52
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at async Microsoft.Azure.WebJobs.Host.Executors.FunctionExecutor.InvokeWithTimeoutAsync(IFunctionInvoker invoker,ParameterHelper parameterHelper,CancellationTokenSource timeoutTokenSource,CancellationTokenSource functionCancellationTokenSource,Boolean throwOnTimeout,TimeSpan timerInterval,IFunctionInstance instance) at C:\projects\azure-webjobs-sdk-rqm4t\src\Microsoft.Azure.WebJobs.Host\Executors\FunctionExecutor.cs : 572
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at async Microsoft.Azure.WebJobs.Host.Executors.FunctionExecutor.ExecuteWithWatchersAsync(IFunctionInstanceEx instance,ParameterHelper parameterHelper,ILogger logger,CancellationTokenSource functionCancellationTokenSource) at C:\projects\azure-webjobs-sdk-rqm4t\src\Microsoft.Azure.WebJobs.Host\Executors\FunctionExecutor.cs : 518
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at async Microsoft.Azure.WebJobs.Host.Executors.FunctionExecutor.ExecuteWithLoggingAsync(IFunctionInstanceEx instance,FunctionStartedMessage message,FunctionInstanceLogEntry instanceLogEntry,ParameterHelper parameterHelper,ILogger logger,CancellationToken cancellationToken) at C:\projects\azure-webjobs-sdk-rqm4t\src\Microsoft.Azure.WebJobs.Host\Executors\FunctionExecutor.cs : 296
   End of inner exception
   at Microsoft.Azure.WebJobs.ServiceBus.SessionMessageProcessor.CompleteProcessingMessageAsync(IMessageSession session,Message message,FunctionResult result,CancellationToken cancellationToken)
   at async Microsoft.Azure.WebJobs.ServiceBus.Listeners.ServiceBusListener.ProcessSessionMessageAsync(IMessageSession session,Message message,CancellationToken cancellationToken)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at async Microsoft.Azure.ServiceBus.SessionReceivePump.MessagePumpTaskAsync(IMessageSession session)`

The combination of these 2 issues causes us to have a huge amount of active and dead lettered messages in our subscription. Every message abandoned seemed to be retried even though the AbandonAsync threw SessionLockLostException. And it is retried for 10 times until it goes to DLQ.

How can I investigate why my function fails to start the orchestrator? I don't see any exception except the orchestator is being pending for a long time (hours).

How long is session lock active? I tried adding sessionHandler.maxAutoRenewDuration but I saw a couple of issues in Azure that says that the setting is not working, like this: https://github.com/Azure/azure-functions-servicebus-extension/issues/144

What do we need to do when handling SessionLockLostException?

Thanks

**Comment by olitomlinson** [comment]
@adeliab 

Hey, I can't provide a solution to your problem, however, I can recall a handful of times where people have tried to do interesting things with the Message lock, in the same Function that starts the Orchestration. Unfortunately, often with little success.

Generally the advice is so not try to blur the Service Bus message lifetime with the Orchestration lifetime. Put simply - If you can _start_ the Orchestration successfully, then complete the SB message asap.

Can I ask the reason why you want to observe the status of the Orchestration and subsequently Abandon the SB message if the Orchestration is not in your desired state? I might be able to provide some other guidance.

**Comment by cgillum** [comment]
As for why the orchestrator functions aren't starting, try using the [Azure Functions Diagnose and Solve tool](https://docs.microsoft.com/en-us/azure/azure-functions/functions-diagnostics) and searching for "Durable Functions" to see if any known issues are being detected. Note that this tool works best when you're using the latest version(s) of the Durable Functions extension.

**Reply by cgillum** [reply]
@adeliab it's a little hard to respond because there are two different issues going on. I'm not an expert on Service Bus and this isn't a good forum for Service Bus-specific questions, but I think the reason you're seeing `SessionLockLostException` is because you're waiting too long to abandon the message. In other words, by the time you're abandoning the message, the session had already expired and was likely picked up by another instance. Increasing the session lock time to be longer than your orchestration wait time or decreasing the wait time to be less than the session lock timeout is how you can fix the problem with dead-lettered messages.

My suggestion about using the Diagnose and Solve tool was not meant to provide insights to the Service Bus behavior, but rather help explain why the orchestrations are not starting. For that issue, I suggest opening a GitHub issue and filling out the template so we can take a closer look. Alternatively, you may want to consider opening a support request (it's more work for you and might take a bit longer, but it at least ensures that your request for help won't get lost).

**Reply by adeliab** [reply]
@cgillum thanks, i created the issue but haven't received a response: https://github.com/Azure/azure-sdk-for-net/issues/25130
indeed i'm thinking of increasing the session timeout, but it's unclear for me how i can do that in azure function. i saw in a couple of threads that the maxAutoRenew in sessionHandler setting doesn't work.
i will see what the team replies in the issue.. thanks in the mean time

**Comment by adeliab** [comment]
@olitomlinson if the orchestrator takes too long (longer than 1 min), we want to abandon the message so the message will go back to the queue. Maybe we should also terminate the orchestrator instance in that case?

@cgillum the exception doesn't seem to occur in the function itself, so it's not detected by this tool

**Comment by olitomlinson** [comment]
@adeliab 

Yes you could terminate the Orchestration, however bare in mind that any Activities that have already been scheduled by the Orchestration instance will continue to run to completion/failure. In real terms, what this means is that just because you terminated your Orchestration after 60 seconds, there may still be processing occurring, which may or may not be a problem in your use-case.

May I ask a question? When 60 seconds elapses, and the message becomes available again on the queue, what then? Are you intending for service bus to redeliver the message and try the whole Orchestration operation again?

If so, can I make a suggestion that you shift this retry logic and 60sec timeout logic from ServiceBus/ Service Bus triggered function, and into the Orchestration itself? An Orchestration is a perfect place to elegantly handle timeouts and subsequent retries of an operation, and this will massively simplify the architecture.

If there are reasons why you must use Service Bus for redelivery on timeout of the orchestration, I would love to hear more to help understand the use-case :)

**Reply by adeliab** [reply]
@olitomlinson thank you for the suggestion - it's a good question, i 'inherited' this code recently and now i'm wondering why they decided to abandon the message but let the orchestration continues to run. i need to rethink the bigger picture of failure handling for these functions.

i'm trying to understand how session lock works though but i find the documentation really unclear. i can't figure out how long is a session lock is, if it's influenced by the message lock, how can i extend it, etc. if you have a better understanding of it, i'd be glad to hear more about it.

also it is really puzzling how suddenly the functions takes longer than a minute to start the orchestrator. we tried a couple of things - code changes, clearing the whole azure storage. it sometimes works and then it comes back again. so the orchestrator is scheduled but it just won't start. do you have any experience with this issue?

**Reply by olitomlinson** [reply]
@adeliab no problem, happy to try and help! Unfortunately, I can't help much on the lock handling stuff, I've steered clear of it as much as possible as it is complex thing, and as I get older my appetite for dealing with complexity is diminishing. hah.

The only thing I can suggest is ripping out the `session` aspect of your process, and see if you can get it working with a normal message first, and that might narrow down the problem somewhat...?

**Going off on a slight detour....**

When it comes to failure handling of queue-triggered functions, its very common to use App Insights (or similar) for tracking the length of the dead letter queue for monitoring and alerting promises. 

This has generated a bit of a conflict when it comes to people adopting Durable Functions, because they tend to apply their general failure handling principles onto Orchestrations i.e. trying to manually complete/abandon the SB message by observing the runtime status of the Orchestration. 

Unfortunately, I've not seen this done successfully yet. Not saying it can't be done, but it involves manual lock handling which in my experience is another level of complexity, and not straight forward. Its not a path well-trodden from my view.

Because of this I've tended to steer clear of doing manual lock handling in Function triggers, and just accepted that once an orchestration is _started_  using `client.StartNewAsync() `then the originating SB message should be auto-completed / completed asap. 

This meant that I had to adjust my approach to failure handling as I could no longer rely on my previous established practice of monitoring deadletter length. 

If you're interested in what I did see this https://github.com/Azure/azure-functions-durable-extension/discussions/1766

p.s. here is some more discussion from not too long ago on a very similar problem with manual lock handling. https://github.com/Azure/azure-functions-durable-extension/discussions/1888

**Reply by adeliab** [reply]
@olitomlinson i see.. so you subscribe to the failed events and replay the payload/ orchestrator as necessary..
in my case the tricky part is ensuring the order of the message. the reason of the message being abandoned is that so it goes back into the queue and the next message for the same session id won't be processed yet. completing the message immediately will cause the next message in the session to be processed

**Reply by olitomlinson** [reply]
Ok that complicates things a touch.

So when the next ordered message comes through, I'm assuming will you instantiate new unique Orchestration? Will this new Orchestration carry over any state from the previous Orchestration (that was triggered from the previous SB message) ?

**Reply by adeliab** [reply]
yes a new orchestration, the previous orchestration state won't be used. but we do modify a durable entity in the orchestration whose state is also being checked by the same orchestration in the process

and these new messages come in fast, not like sporadically

**Reply by olitomlinson** [reply]
Ok, so does the Orchestration do anything, such as call Activities before/after it calls/signals to the Entity? Or does the orchestration only exist to facilitate a call to the Entity, and nothing else?

**Reply by adeliab** [reply]
It differs per orchestrators but it calls activity/ sub orchestrators

**Reply by olitomlinson** [reply]
Ok. So I think there is a pattern for this (I'm doing something similar here https://github.com/Azure/azure-functions-durable-extension/discussions/1980 - just waiting on some guidance from the team on how well it will scale)

The idea is that you have an Orchestration that is just waiting for the same type of external events (in an indefinite loop), once that event arrives, "do something with it", and then begin waiting for that event happen again, _Ad Infinitum_

in your case "do something with it'" is calling your existing Orchestration process.

I believe events raised using the `RaiseEventAsync` API are handled in-order by the Orchestration in the order that they are published. 

**some very rough pseudo...**

```c#
public async Task Run([ServiceBusTrigger("mySessionQueue", AccessRights.Manage)] Message myMessage , [DurableClient] IDurableOrchestrationClient client,)
{
   var orchestration  = (get orchestration where Id is 'SessionA')
   
   if (orchestration is null | orchestration is not running)
   {
       orchestration = (Start orchestration with id = 'Session A')
   }
   
   await client.RaiseEventAsync("Session A", "CertainEvent", myMessage);
}
```



**

**Reply by adeliab** [reply]
thanks for the example, but i don't really get the idea..

so let's say we have session A with message 1 and message 2.
message 1 comes, the function is triggered. it then gets orchestration (?) - what is this get, get by what?
the function then creates an orchestration - so just schedule without starting? which method do i need to use for this?
then we call the raise event, and we only start the orchestrator in the event listener?

so let's say after this the message is completed and message 2 comes. what if message 1 orchestrator fails? message 2 will still be processed then

**Reply by olitomlinson** [reply]
I've updated the example, with the notion of "Session A" - that might make things clearer?

The service bus trigger function will keep pushing your messages (`myMessage`), in order, to the Session A orchestration, as they happen.

These events will queued up, in-order, in the Runtime waiting to be dequeued by the Orchestration.

Now skip to my example here https://github.com/Azure/azure-functions-durable-extension/discussions/1980 and imagine this is the inside of your Orchestration that is handling "Session A"

notice the while loop, notice that the function is continually waiting and dequeuing the same event type indefinitely. 

```c#
while (true) {
   var registration = context.WaitForExternalEvent<string>(Orchestrations.REGISTER_ATTENDEE);
   
   // Handle the event, call other sub-orchestrations etc until you are ready to continue around the loop and grab the next event
}

```
---

Basically all I'm doing here is utilising the Runtime as a FIFO queue to push work into the Orchestration, and letting the orchestration naturally attempt to drain that queue by rescheduling `context.WaitForExternalEventAsync` as soon as an event has been handled.

What I'm unclear on is how well this model scales, does it start shedding messages if they've not been processed in so many itterations etc - which is the purpose of my question in  https://github.com/Azure/azure-functions-durable-extension/discussions/1980

**Reply by olitomlinson** [reply]
Okay so I've got some more clarity on #1980 and so what I'm recommending to you may not work... hold fire!

**Reply by olitomlinson** [reply]
Yep, unfortunately, this won't work because the runtime won't queue the results to be used in a FIFO way like I assumed it would. It actually uses a Stack. So there is a potential for this to mess up the ordering in a high frequency scenario, which it sounds like your use-case is.

There is another solution to use a Durable Entity which performs the role of a FIFO queue which I'm confident would work, but at this point it feels like we might be hacking a little too far around the actual issues...

It's worth following Chris's advice here for next steps https://github.com/Azure/azure-functions-durable-extension/discussions/1985#discussioncomment-1596873

If you don't get anywhere there, happy to help you on the other solution I just mentioned :)

**Reply by adeliab** [reply]
many thanks for thinking along with this issue @olitomlinson  i've created an issue to see what MS suggests in this scenario
